Add pattern to detect for loop indexing into tensor

davidchencsl opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81740
* #81733
* #81501
* #81273
* __->__ #81056

Differential Revision: [D37725266](https://our.internmc.facebook.com/intern/diff/D37725266)

Summary: This commit added a pattern in the pattern matcher to detect if for loop & indexing is used that can be vectorized.

Test Plan: Added a test in test_profiler.py that creates a bootstrap example.

davidchencsl committed:
Add pattern to detect for loop indexing into tensor

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/init.cpp:
```diff
@@ -258,12 +258,22 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject* unused) {
 
   {
     using torch::profiler::impl::Result;
+    py::class_<PyFrameState>(m, "_PyFrameState")
```
I think this is ok to expose some of these objects but maybe these should be in their own submodule and not in the autograd/profiler bindings?

robieta replied:
Yeah, it's probably about time to make a profiler submodule. Those are in autograd mostly out of inertia.

davidchencsl committed:
Update on "Add pattern to detect for loop indexing into tensor"




[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37725266).

davidchencsl committed:
Update on "Add pattern to detect for loop indexing into tensor"


Differential Revision: [D37725266](https://our.internmc.facebook.com/intern/diff/D37725266)

Summary: This commit added a pattern in the pattern matcher to detect if for loop & indexing is used that can be vectorized.

Test Plan: Added a test in test_profiler.py that creates a bootstrap example.

[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37725266).

robieta reviewed (commented):
A couple minor nits, but otherwise looks good.

davidchencsl commented on torch/profiler/_pattern_matcher.py:
```diff
@@ -33,8 +40,10 @@ def match(self, event: _ProfilerEvent):
+            return []
         matched_events = []
-        for event in eventTreeDFS(self.event_tree):
+        for event in eventTreeBFS(self.event_tree):
```
Its because I need to have access to the events in the same level in the order as they take place. But you could also argue that I could just reverse the order that I add children into stack. What do you think makes more sense?

robieta replied:
I thought the semantics of `match` were that you could traverse the events in any order?

robieta replied:
Maybe this is a sign that match should be matching a higher level node?

davidchencsl replied:
Uh yes, but the time complexity would be too high if i don't keep a visited set and avoid. redundant visited. For example, one aten::select node in a for loop implies that all the aten::select in the for could be a match. By keeping a visited set I can avoid revisited these. But that requires me to traverse in the particular order..

robieta replied:
In that case it should probably be part of the API so the matcher can specify the traversal order that it expects.

davidchencsl committed:
Update on "Add pattern to detect for loop indexing into tensor"


Differential Revision: [D37725266](https://our.internmc.facebook.com/intern/diff/D37725266)

Summary: This commit added a pattern in the pattern matcher to detect if for loop & indexing is used that can be vectorized.

Test Plan: Added a test in test_profiler.py that creates a bootstrap example.

[ghstack-poisoned]

davidchencsl committed:
Update on "Add pattern to detect for loop indexing into tensor"


Differential Revision: [D37725266](https://our.internmc.facebook.com/intern/diff/D37725266)

Summary: This commit added a pattern in the pattern matcher to detect if for loop & indexing is used that can be vectorized.

Test Plan: Added a test in test_profiler.py that creates a bootstrap example.

[ghstack-poisoned]

davidchencsl committed:
Update on "Add pattern to detect for loop indexing into tensor"


Differential Revision: [D37725266](https://our.internmc.facebook.com/intern/diff/D37725266)

Summary: This commit added a pattern in the pattern matcher to detect if for loop & indexing is used that can be vectorized.

Test Plan: Added a test in test_profiler.py that creates a bootstrap example.

[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37725266).

davidchencsl commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot revert -m "broke 10.2 test_forward_ad_nn_functional_conv_transpose3d_cuda_float32 test on trunk https://hud.pytorch.org/pytorch/pytorch/commit/6e621d028766808f3c0bc8c45f776af3bf9cf22f" -c nosignal

```
======================================================================
2022-07-20T20:45:44.3845308Z FAIL [2.091s]: test_forward_ad_nn_functional_conv_transpose3d_cuda_float32 (__main__.TestCompositeComplianceCUDA)
2022-07-20T20:45:44.3848588Z ----------------------------------------------------------------------
2022-07-20T20:45:44.3849171Z Traceback (most recent call last):
2022-07-20T20:45:44.3849722Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1883, in wrapper
2022-07-20T20:45:44.3850102Z     method(*args, **kwargs)
2022-07-20T20:45:44.3850583Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1883, in wrapper
2022-07-20T20:45:44.3850953Z     method(*args, **kwargs)
2022-07-20T20:45:44.3853385Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
2022-07-20T20:45:44.3853810Z     result = test(self, **param_kwargs)
2022-07-20T20:45:44.3854428Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 813, in test_wrapper
2022-07-20T20:45:44.3867042Z     return test(*args, **kwargs)
2022-07-20T20:45:44.3867447Z   File "/var/lib/jenkins/workspace/test/test_ops.py", line 1258, in test_forward_ad
2022-07-20T20:45:44.3867861Z     composite_compliance.check_forward_ad_formula(
2022-07-20T20:45:44.3868483Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/composite_compliance.py", line 567, in check_forward_ad_formula
2022-07-20T20:45:44.3868974Z     assert_equal_fn(actual_primals, expected_primals, equal_nan=True)
2022-07-20T20:45:44.3869572Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 2335, in assertEqual
2022-07-20T20:45:44.3869951Z     assert_equal(
2022-07-20T20:45:44.3870410Z   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
2022-07-20T20:45:44.3870801Z     raise error_metas[0].to_error(msg)
2022-07-20T20:45:44.3871179Z AssertionError: Tensor-likes are not close!
2022-07-20T20:45:44.3871391Z 
2022-07-20T20:45:44.3871526Z Mismatched elements: 1 / 2744 (0.0%)
2022-07-20T20:45:44.3871974Z Greatest absolute difference: 4.57763671875e-05 at index (2, 3, 4, 4) (up to 2e-05 allowed)
2022-07-20T20:45:44.3872504Z Greatest relative difference: 5.68618647611624e-05 at index (2, 3, 4, 4) (up to 2e-05 allowed)
2022-07-20T20:45:44.3872743Z 
2022-07-20T20:45:44.3873013Z ----------------------------------------------------------------------
```

janeyx99 commented:
LOL nice the revert didn't go through because it was going to be a false revert--the job was green again after, so I suppose it was a flaky error of some sort. Sorry about the extra noise!

davidchencsl commented:
Oh my, looks like it does broke a test outside of CI: https://github.com/pytorch/pytorch/pull/81812
